src/main/java/org/radarcns/service/DataSetService.java

```diff
@@ -197,4 +241,66 @@ private Header getHeader(String project, String subject, String source,
         return new Header(project, subject, source, sourceType, sourceData.getSourceDataType(),
                 stat, sourceData.getUnit(), timeWindow, timeFrame, null);
     }
+
+    /**
+     * Returns calculated {@link AggregatedDataPoints} using given parameters.
+     *
+     * @param projectName of project
+     * @param subjectId of subject
+     * @param sources requested
+     * @param timeWindow interval
+     * @param start time
+     * @param end time
+     * @return calculated data.
+     */
+    public AggregatedDataPoints getAggregatedData(String projectName, String subjectId,
+            List<AggregateDataSource> sources, TimeWindow timeWindow, Instant start, Instant end) {
+        List<DataItem> dataItems = calculateIntervals(start, end, timeWindow).stream()
+                .map(p -> computeAggregatedDataItem(projectName, subjectId, sources, p, timeWindow))
+                .collect(Collectors.toList());
+        // fill up sourceData.type field
+        sources.forEach(d -> d.getSourceData().forEach(e -> {
+            try {
+                e.setType(sourceCatalog
+                        .getSourceData(e.getName()).getSourceDataType());
+            } catch (IOException exe) {
+                throw new BadGatewayException(exe);
+            }
+        }));
+
+        return new AggregatedDataPoints(projectName, subjectId,
+                dataItems.stream().map(p -> (Integer) p.getSample()).reduce(Integer::max).get(),
+                new TimeFrame(start, end), timeWindow, sources, dataItems);
+    }
+
+    private DataItem computeAggregatedDataItem(String projectName, String subjectId,
+            List<AggregateDataSource> aggregateDataSources, TimeFrame timeFrame,
+            TimeWindow timeWindow) {
+        Integer count = 0;
+        for (AggregateDataSource aggregate : aggregateDataSources) {
```

Comment:
You could also use a nested `stream().reduce()` here, but this approach is fine as well.